synapseclient/producer_consumer_download_threads.py

```diff
@@ -0,0 +1,282 @@
+                    printTransferProgress(transferred, to_be_transferred, 'Downloading ',
+                                          os.path.basename(path), dt=time.time()-t_first_write)
+            finally:
+                self.data_queue.task_done()
+
+
+def download_files(client, download_requests: typing.Sequence[DownloadRequest], num_threads):
+    """
+    Main driver for the multi-threaded download. Uses the producer-consumer with Queue design pattern as described
+    in Effective Python Item 39.
+
+    :param client: A SynapseBaseClient
+    :param download_requests: A batch of DownloadRequest objects specifying what Synapse files to download
+    :param num_threads: The number of download threads
+    :return: Map between each DownloadRequest in download_requests object and the corresponding DownloadResponse object
+    """
+
+    session = requests.Session()
+    retry = Retry(connect=CONNECT_FACTOR, backoff_factor=BACK_OFF_FACTOR)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount('http://', adapter)
+    session.mount('https://', adapter)
+
+    data_queue = Queue(BUF_SIZE)
+    range_queue = Queue(BUF_SIZE)
+    producer_threads = []
+    consumer_threads = []
+
+    for request in download_requests:
+        producer_threads_local = []
+        consumer_threads_local = []
+        for _ in range(num_threads):
+            producer_worker = ProducerDownloadThread(client, request, range_queue, data_queue)
+            consumer_worker = ConsumerDownloadThread(data_queue)
+            producer_threads_local.append(producer_worker)
+            consumer_threads_local.append(consumer_worker)
+
+        for producer_thread, consumer_thread in zip(producer_threads_local, consumer_threads_local):
+            producer_thread.start()
+            consumer_thread.start()
+
+        for producer_thread, consumer_thread in zip(producer_threads_local, consumer_threads_local):
+            producer_threads.append(producer_thread)
+            consumer_threads.append(consumer_thread)
+
+        global transferred
+        global t_first_write
+        global previously_transferred
+        transferred = t_first_write = previously_transferred = 0
+
+        file_name, pre_signed_url = _get_pre_signed_batch_request_json(client, request)
+        res_get = session.get(pre_signed_url, stream=True)
+        file_size = int(res_get.headers['Content-Length'])
+        pre_signed_url_chunk_generator = _get_chunk_pre_signed_url(file_size,
+                                                                   file_name,
+                                                                   pre_signed_url,
+                                                                   request.path)
+        global to_be_transferred
+        to_be_transferred = file_size
+        _create_empty_file(file_size, request.path)
+        for chunk in pre_signed_url_chunk_generator:
+            range_queue.put(chunk)
+    range_queue.join()
```

Comment:
Could this allow consumers to terminate if producer is too slow to populate consumer queue, i.e. what is termination condition for consumer queue?

Reply:
Added a CloseableQueue class which allows producers to signal when they are done producing, this way consumers will terminate only when there is no more data yet to be produced.